
# wrnchAI Hand's on

## Objectives

At the end of this presentation you should be able install and use wrnchAI

- What's wrnch?
    - Live example
    - Introduction to pose estimation
    - Joints definition

- Getting started with:
    - Cloud REST API
    - Edge SDK

This getting started notebook will be shared to all participants

# Whats wrnch?

## Human pose estimation
- 2d, 3d, head and hand pose estimations
- Activity recognition (with Edge SDK)

## How to use wrnch
- Cloud: REST API
- Edge: `Python` & `C++` API 

---
# Getting started with joint definition 

## 1: Register and test wrnchAI
### Register you developer account

- Register throught **this link to get a 5 days free dev license** : https://devportal.wrnch.ai/codejam2019
- NB: if you don't use this URL, you **won't** receive a free license
    



### Submit an image or a video
- For this demo we will use this image:


![test-image](img/small/test-image.jpg)


- Once register you can submit video to our test page to this address:
    https://devportal.wrnch.ai/wrnchcloud

- Ask for both JSON and Annotated media with those options:
    - Head Pose Estimation
    - Hand Pose Estimation

# 2: Retrive results

Once our servers has processed the image, download the results. If you have selected more than one worktype, you will recieve them as a zip. In our example the zip has an annotated_media and a json file

The annotated media will look like this:

![result img](img/small/test-image-processed.jpeg)


# 3. Process the results

## Our goal
-> Lets verify that his left wrist is on his left knee

### 1. Import the json file
 

In [1]:
import json

def json_r(filename):
   with open(filename) as f_in:
       return(json.load(f_in))

pose_estimation = json_r('test-image-processed.json')

print(str(pose_estimation['frames'])[:300])

[{'frame_time': 0, 'height': 819, 'persons': [{'hand_pose': {'left': {'bbox': {'height': 0.2857142984867096, 'minX': 0.6681647300720215, 'minY': 0.3588320314884186, 'width': 0.29249998927116394}, 'is_main': False, 'joints': [0.7872003316879272, 0.49528399109840393, 0.7665179371833801, 0.510883808135


In [2]:
man_pose = pose_estimation['frames'][0]['persons'][0]

print(man_pose.keys())

dict_keys(['hand_pose', 'head_pose', 'id', 'pose2d'])


In [3]:
import pprint
# Lets print the 10 first joint coordinates
pprint.pprint(man_pose['pose2d']['joints'][0:10])

[0.3711889684200287,
 0.7936080694198608,
 0.5704107880592346,
 0.7095877528190613,
 0.3788127601146698,
 0.6029024720191956,
 0.5351067781448364,
 0.5762162208557129,
 0.7694329023361206,
 0.5570101141929626]


### 2. Interpret the coordinates
 
 Each joint has a 2d coordinate in this space: 
 
 ![joint coor](img/small/with-arrow.png)
 
 #### We define both 2d and 3d coordinate on this page: https://devportal.wrnch.ai/wrnchai_sdk/coord_spaces

### Joints definition
#### For more information and figure, refer to our joint definition documentation: 
https://devportal.wrnch.ai/wrnchai_sdk/joint_definitions


![hands joint doc](img/small/j25_body_joints.jpg)


#### There is 25 joints. The json should have 50 coordinates

In [4]:
print(len(man_pose['pose2d']['joints']))

50


Lets determine if his left wrist is over his left knee.
We should have :

`xLWRIST ~= xLKNEE`

In [5]:
xLWRIST = man_pose['pose2d']['joints'][30]
xLKNEE = man_pose['pose2d']['joints'][8]

print(xLWRIST)
print(xLKNEE)

0.7735595107078552
0.7694329023361206


-> Indeed the wrist and the elbow are close is the frame

---

# Getting started with the cloud API
All of this is also documented here: https://devportal.wrnch.ai/wrnchcloud/api_docs#introduction

First install requests. The python http client.

In [6]:
!pip install requests
import requests
import json

You are using pip version 19.0.3, however version 19.3.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


The two paths we will need for this example

In [7]:
LOGIN_URL = 'https://api.wrnch.ai/v1/login'
JOBS_URL = 'https://api.wrnch.ai/v1/jobs'

1. Retrieve the cloud API key

Click on the key icon on this page: https://devportal.wrnch.ai/licenses to find your cloud API key

In [15]:
#Save the Cloud API key for next cell to work
API_KEY = ""

2. Authenticate

In [9]:
resp_auth = requests.post(LOGIN_URL,data={'api_key':API_KEY})
print(resp_auth.text)
# the jwt token is valid for an hour
JWT_TOKEN = json.loads(resp_auth.text)['access_token']

{"access_token": "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOjc1NiwibGljZW5zZV9pZCI6NTkxLCJhZGRvbnMiOnt9LCJleHAiOjE1NzM4MzkyNjMsImlkZW50aXR5Ijo3NTYsInJvbGUiOiJpbmRfY2xpZW50IiwiaWF0IjoxNTczODMyMDYzLCJqdGkiOiJhNzYyYzBiYWVmZmVmMmUzY2Y1ZTllMmM5Y2YxNjdhNDE5M2I5YzcxMDBiZTUyNDhmMzcyZDBiNzQxNWVjNDMwIiwidHlwZSI6ImFjY2VzcyIsImZyZXNoIjoiZmFmIn0.OOWitopW0bsZ0Ne5hLsyhh3PzctKvBm_HLwUeYYXLQ4"}



Now lets submit a file for processing
1. Open the file as a byte stream
2. Send a post request with authentification and the file

In [10]:
with open('img/test-image.jpg', 'rb') as f:
    resp_sub_job = requests.post(JOBS_URL,
                                 headers={'Authorization':f'Bearer {JWT_TOKEN}'},
                                 files={'media':f},
                                 data={'work_type':'json'}
                                )

In [11]:
job_id = json.loads(resp_sub_job.text)['job_id']
print('Status code:',resp_sub_job.status_code)
print('Response:',resp_sub_job.text)

Status code: 202
Response: {"job_id": "1cc5c026-1ac8-483b-8ff5-83ec09fa425e"}



The status code should be 202 and return a job_id.

Now lets retrieve the result

In [12]:
GET_JOB_URL = JOBS_URL + '/' +job_id
print(GET_JOB_URL)
resp_get_job = requests.get(GET_JOB_URL,headers={'Authorization':f'Bearer {JWT_TOKEN}'})
print('Status code:',resp_get_job.status_code)
print('\nResponse:',resp_get_job.text)

https://api.wrnch.ai/v1/jobs/1cc5c026-1ac8-483b-8ff5-83ec09fa425e
Status code: 200

Response: {"file_info":{"joint_definitions":{"hands":"hand21","head":"wrFace20","pose2d":"j25","pose3d_ik":"extended","pose3d_raw":"j25"}},"frames":[{"frame_time":0,"height":614,"persons":[{"id":0,"pose2d":{"bbox":{"height":0.9895586371421814,"minX":0.2897970974445343,"minY":0.039630815386772156,"width":0.5532262325286865},"is_main":true,"joints":[0.37112537026405334,0.7902586460113525,0.5703703165054321,0.7099169492721558,0.3749939203262329,0.6030898690223694,0.5311734676361084,0.5763961672782898,0.7657291889190674,0.5610739588737488,0.6757756471633911,0.866474986076355,0.45308369398117065,0.5897430181503296,0.5196003913879395,0.2996460795402527,0.5196273922920227,0.29780837893486023,0.5455899834632874,0.13356269896030426,0.5350296497344971,0.5954576730728149,0.3633612096309662,0.46940380334854126,0.42190080881118774,0.28249606490135193,0.6172999143600464,0.31679609417915344,0.6875200271606445,0.458121

Lets see if the left wrist and left knee are touching:

In [13]:
cloud_pose_estimation = json.loads(resp_get_job.text)
cloud_xLWRIST = cloud_pose_estimation['frames'][0]['persons'][0]['pose2d']['joints'][30]
cloud_xLKNEE = cloud_pose_estimation['frames'][0]['persons'][0]['pose2d']['joints'][8]

print(cloud_xLWRIST)
print(cloud_xLKNEE)

0.7735621333122253
0.7657291889190674


-> Now you can use human pose estimation in your project with very few line of code

---

# Getting started with the edge sdk

The edge skd allow you to do real time pose estimation.

For installation procedure refer to the detail documentation here: 
https://devportal.wrnch.ai/wrnchai_sdk/releases

NB: We do not support MacOS yet.